README.rst
OHC - An off-heap-cache
=======================

Features
========

- asynchronous cache loader support
- optional per entry or default TTL/expireAt
- entry eviction and expiration without a separate thread
- capable of maintaining huge amounts of cache memory
- suitable for tiny/small entries with low overhead using the chunked implementation
- runs with Java 8 and Java 11 - support for Java 7 and earlier has been dropped with version 0.7.0
- to build OHC from source, Java 11 or newer (tested with Java 11 + 15) is required

Performance
===========

OHC shall provide a good performance on both commodity hardware and big systems using non-uniform-memory-architectures.

No performance test results available yet - you may try the ohc-benchmark tool. See instructions below.
A very basic impression on the speed is in the _Benchmarking_ section.

Requirements
============

Java 8 VM that support 64bit and has ``sun.misc.Unsafe`` (Oracle JVMs on x64 Intel CPUs).

OHC is targeted for Linux and OSX. It *should* work on Windows and other Unix OSs.

Architecture
============

OHC provides two implementations for different cache entry characteristics:
- The _linked_ implementation allocates off-heap memory for each entry individually and works best for medium and big entries.
- The _chunked_ implementation allocates off-heap memory for each hash segment as a whole and is intended for small entries.

Linked implementation
---------------------

The number of segments is configured via ``org.caffinitas.ohc.OHCacheBuilder``, defaults to ``# of cpus * 2`` and must
be a power of 2. Entries are distribtued over the segments using the most significant bits of the 64 bit hash code.
Accesses on each segment are synchronized.

Each hash-map entry is allocated individually. Entries are free'd (deallocated), when they are no longer referenced by
the off-heap map itself or any external reference like ``org.caffinitas.ohc.DirectValueAccess`` or a
``org.caffinitas.ohc.CacheSerializer``.

The design of this implementation reduces the locked time of a segment to a very short time. Put/replace operations
allocate memory first, call the ``org.caffinitas.ohc.CacheSerializer`` to serialize the key and value and then put the
fully prepared entry into the segment.

Eviction is performed using an LRU algorithm. A linked list through all cached elements per segment is used to keep
track of the eldest entries.

Chunked implementation
----------------------

Chunked memory allocation off-heap implementation.

Purpose of this implementation is to reduce the overhead for relatively small cache entries compared to the linked
implementation since the memory for the whole segment is pre-allocated. This implementation is suitable for small
entries with fast (de)serialization implementations of ``org.caffinitas.ohc.CacheSerializer``.

Segmentation is the same as in the linked implementation. The number of segments is configured via
``org.caffinitas.ohc.OHCacheBuilder``, defaults to ``# of cpus * 2`` and must be a power of 2. Entries are distribtued
over the segments using the most significant bits of the 64 bit hash code. Accesses on each segment are synchronized.

Each segment is divided into multiple chunks. Each segment is responsible for a portion of the total capacity
``(capacity / segmentCount)``. This amount of memory is allocated once up-front during initialization and logically
divided into a configurable number of chunks. The size of each chunk is configured using the ``chunkSize`` option in
``org.caffinitas.ohc.OHCacheBuilder``.

Like the linked implementation, hash entries are serialized into a temporary buffer first, before the actual put
into a segment occurs (segement operations are synchronized).

New entries are placed into the current write chunk. When that chunk is full, the next empty chunk will become the new
write chunk. When all chunks are full, the least recently used chunk, including all the entries it contains, is evicted.

Specifying the ``fixedKeyLength`` and ``fixedValueLength`` builder properties reduces the memory footprint by
8 bytes per entry.

Serialization, direct access and get-with-loader functions are not supported in this implementation.

To enable the chunked implementation, specify the ``chunkSize`` in ``org.caffinitas.ohc.OHCacheBuilder``.

Note: the chunked implementation should still be considered experimental.

Eviction algorithms
===================

OHC supports three eviction algorithms:

- *LRU*: The oldest (least recently used) entries are evicted to make room for new entries.
- *Window Tiny-LFU*:
  Entries with lower usage frequency are evicted to make room for new entries.
  The goal of this eviction algorithm is to prevent heavily used entries from being evicted.
  Note that the maximum size of entries is limited to the size of the eden generation, which is currently
  fixed at 20% of the segment size (i.e. overall capacity / number of segments).
  Each OHC cache segment is divided into an eden and a main "generation". New entries start in the eden generation
  to give these time to build up their usage frequencies. When the eden generation becomes full, entries in the
  eden generation have to pass the admission filter, which checks the frequencies of the entries in the eden
  generation against the frequencies of the oldest (least recently used) entries in the main generation.
  See `this article <http://highscalability.com/blog/2016/1/25/design-of-a-modern-cache.html>`_ for a more thorough
  description.
  (Only supported in the _linked_ implementation, not supported by the chunked implementation)
- *None*: OHC performs no eviction on its own. It is up to the caller to check the return values and monitor
  free capacity.
  (Only supported in the _linked_ implementation, not supported by the chunked implementation)

Configuration
=============

Use the class ``OHCacheBuilder`` to configure all necessary parameter like

- number of segments (must be a power of 2), defaults to number-of-cores * 2
- hash table size (must be a power of 2), defaults to 8192
- load factor, defaults to .75
- capacity for data over the whole cache
- key and value serializers
- default TTL
- optional unlocked mode

Generally you should work with a large hash table. The larger the hash table, the shorter the linked-list in each
hash partition - that means less linked-link walks and increased performance.

The total amount of required off heap memory is the *total capacity* plus *hash table*. Each hash bucket (currently)
requires 8 bytes - so the formula is ``capacity + segment_count * hash_table_size * 8``.

OHC allocates off-heap memory directly bypassing Java's off-heap memory limitation. This means, that all
memory allocated by OHC is not counted towards ``-XX:maxDirectMemorySize``.

Memory & jemalloc
=================

Since especially the linked implementation performs alloc/free operations for each individual entry, consider that
memory fragmentation can happen.

Also leave some head room since some allocations might still be in flight and also "the other stuff"
(operating system, JVM, etc) need memory. It depends on the usage pattern how much head room is necessary.
Note that the linked implementation allocates memory during write operations _before_ it is counted towards the
segments, which will evict older entries. This means: do not dedicate all available memory to OHC.

We recommend using jemalloc to keep fragmentation low. On Unix operating systems, preload jemalloc.

OSX usually does not require jemalloc for performance reasons. Also make sure that you are using a recent version of
jemalloc - some Linux distributions still provide quite old versions.

To preload jemalloc on Linux, use
``export LD_PRELOAD=<path-to-libjemalloc.so``, to preload jemalloc on OSX, use
``export DYLD_INSERT_LIBRARIES=<path-to-libjemalloc.so``. A script template for preloading can be found at the
`Apache Cassandra project <https://github.com/apache/cassandra/blob/bf3255fc93db65b816b016958967003df38a6004/bin/cassandra#L135-L182>`_.

Usage
=====

Quickstart::

 OHCache ohCache = OHCacheBuilder.newBuilder()
                                 .keySerializer(yourKeySerializer)
                                 .valueSerializer(yourValueSerializer)
                                 .build();

This quickstart uses the very least default configuration:

- total cache capacity of 64MB or 16 * number-of-cpus, whichever is smaller
- number of segments is 2 * number of cores
- 8192 buckets per segment
- load factor of .75
- your custom key serializer
- your custom value serializer
- no maximum serialized cache entry size

See javadoc of ``CacheBuilder`` for a complete list of options.

Key and value serializers need to implement the ``CacheSerializer`` interface. This interface has three methods:

- ``int serializedSize(T t)`` to return the serialized size of the given object
- ``void serialize(Object obj, DataOutput out)`` to serialize the given object to the data output
- ``T deserialize(DataInput in)`` to deserialize an object from the data input

Building from source
====================

Clone the git repo to your local machine. Either use the stable master branch or a release tag.

``git clone https://github.com/snazy/ohc.git``

You need OpenJDK 11 or newer to build from source. Just execute

``mvn clean install``

Benchmarking
============

You need to build OHC from source because the big benchmark artifacts are not uploaded to Maven Central.

Execute ``java -jar ohc-benchmark/target/ohc-benchmark-0.7.1-SNAPSHOT.jar -h`` (when building from source)
to get some help information.

Generally the benchmark tool starts a bunch of threads and performs _get_ and _put_ operations concurrently
using configurable key distributions for _get_ and _put_ operations. Value size distribution also needs to be configured.

Available command line options::

 -cap <arg>    size of the cache
 -d <arg>      benchmark duration in seconds
 -h            help, print this command
 -lf <arg>     hash table load factor
 -r <arg>      read-write ration (as a double 0..1 representing the chance for a read)
 -rkd <arg>    hot key use distribution - default: uniform(1..10000)
 -sc <arg>     number of segments (number of individual off-heap-maps)
 -t <arg>      threads for execution
 -vs <arg>     value sizes - default: fixed(512)
 -wkd <arg>    hot key use distribution - default: uniform(1..10000)
 -wu <arg>     warm up - <work-secs>,<sleep-secs>
 -z <arg>      hash table size
 -cs <arg>     chunk size - if specified it will use the "chunked" implementation
 -fks <arg>    fixed key size in bytes
 -fvs <arg>    fixed value size in bytes
 -mes <arg>    max entry size in bytes
 -unl          do not use locking - only appropiate for single-threaded mode
 -hm <arg>     hash algorithm to use - MURMUR3, XX, CRC32
 -bh           show bucket historgram in stats
 -kl <arg>     enable bucket histogram. Default: false

Distributions for read keys, write keys and value sizes can be configured using the following functions::

 EXP(min..max)                        An exponential distribution over the range [min..max]
 EXTREME(min..max,shape)              An extreme value (Weibull) distribution over the range [min..max]
 QEXTREME(min..max,shape,quantas)     An extreme value, split into quantas, within which the chance of selection is uniform
 GAUSSIAN(min..max,stdvrng)           A gaussian/normal distribution, where mean=(min+max)/2, and stdev is (mean-min)/stdvrng
 GAUSSIAN(min..max,mean,stdev)        A gaussian/normal distribution, with explicitly defined mean and stdev
 UNIFORM(min..max)                    A uniform distribution over the range [min, max]
 FIXED(val)                           A fixed distribution, always returning the same value

CHANGES.txt
0.7.1
=====

#67 Improve lock performance
- Requires Java 11 or newer to build, runtime support for Java 8 and newer
  (ohc-core.jar is no longer a multi-release jar)
- Simplified test-related parts in pom.xml
- Bump many library versions (see 3f6614707b34aa8689afe069d74ed319c8a8b2ea + e74d45518168710bc33d99f33cc1e5ed735eeb94)
- Fix javadoc + jmh invocations for Java 11+

0.7.0
=====

- introduced experimental support for Java 11
  - ohc-core.jar is now built as a multi-release jar
  - hasher implementation for CRC32C, falls back to CRC32 before Java 11
#53 Fix calculation of 'slots' in Timeouts
#59 Fix initialization of 'precision' in Timeout
#61 fix double counting in Timeouts.removeExpired()
- upgrade a bunch of dependencies
  - Guave to version 19.0
  - SLF4J to 1.7.25
  - snappy to 1.1.7.2
  - jna to 4.5.2
- removed support for Java 7, therefore the artifact ohc-core-j8 is no longer necessary


0.6.1
=====

- update maven plugin versions
- update jmh version
#34 add WTinyLFU
#37 non-evicting mode
- unit test fixes for systems with few CPUs
- unit test fix regarding charset

0.6.0
=====

BREAKING CHANGES
----------------

- If entry TTLs are used, it is required to configure the cache with OHCacheBuilder.timeouts(true)
- Signatures of OHCache.put() and OHCache.remove() methods has been changed to return a boolean value.
  This means that this version of OHC requires a recompilation of code that uses OHC and cannot be
  used as a drop-in replacement (i.e. just replace a jar).

non-breaking changes
--------------------

(#37 non-evicting mode) - not merged by accident
(#34 add WTinyLFU) - not merged by accident
- add new builder cache flag 'timeouts'
- add Ticker interface
(- unit test fixes for systems with few CPUs) - not merged by accident
(- unit test fix regarding charset) - not merged by accident
- update to guava 18.0
#33 crc32 algorithm does not work on Java7

0.5.1
=====

#45 chunked implementation: fix descruction of off-heap-maps on initialization failure
#44 chunked implementation: fix wrong size calculation on eviction
#32 chunked implementation: allow remove() on returned iterators & do not return removed entries

0.5.0
=====

#30 fix hash table size check for extreme hash table sizes
#27 replace ReentrantLock with CAS
#26 add chunked implementation
#13 remove tables implementation

0.4.5
=====

#23 getWithLoader needs contract for null/non-existant values
- extend javadoc to explain API contracts
- fix java.version check for Java 9 EA builds

0.4.4
=====

#20 TTL support for entries
#9 allow unsynchronized access
#19 iterator may keep entries referenced if iterator is closed before exhausted
#17 Checking for java 8 fails

0.4.3
=====

#18 Allow skipping LRU stats updates in getDirect()

0.4.2
=====

- use Java byte order instead of native byte order for keys and values

0.4.0
=====

#15 Refactor serialization to use java.nio.ByteBuffer instead of streams

0.3.6
=====
#11 add putIfAbsentDirect() and addOrReplaceDirect()
#12 check test coverage of direct methods

0.3.5
=====
#10 OHC fails to load on 1.7 Zing VM (also https://issues.apache.org/jira/browse/CASSANDRA-9296)

0.3.4
=====
#5 add CRC32 and xx-hash as alternative hash algorithms
- use Unsafe to allocate DirectByteBuffer
- Update to recent Maven plugins

0.3.3
=====
- remove jemalloc via JNA Library allocator
- make JNA Native.malloc the default memory allocation strategy

0.3.2
=====
#2 implement direct buffer access to cached values
- added some support for unattended batch of benchmarks
- don't print whole exception message if jemalloc cannot be found (only if debug logging is enabled)

0.3.1
=====
- change system property name from DISABLE_JEMALLOC to org.caffinitas.ohc.disableJEmalloc
- change system property name from DEBUG_OFF_HEAP_MEMORY_ACCESS to org.caffinitas.ohc.debugOffHeapAccess
- add option OHCacheBuilder to throw OutOfMemoryError when off-heap memory-allocation fails

0.3
===
- remove cassandra-all dependency from ohc-benchmark
- do not deploy big binary ohc-benchmark and ohc-jmh artifacts to Maven Central
- change scope of snappy dependency in ohc-core to provided
- benchmark: add new command line option to specify key length (-kl)
- fix murmur3 hashing
- divide overall capacity to all segments and manage free capacity inside segments
- allow use of native malloc()/free() functions via JNA Native class
- allow to specify preferred memory allocation via system property "org.caffinitas.ohc.allocator"
- allow to specify defaults of OHCacheBuilder via system properties prefixed with "org.caffinitas.org."
- max entry size segment defaults to capacity per segment
- remove fastutil dependency
#3 support cache-loader functionality
- some dependency cleanups in pom.xml files
- benchmark: make metrics in local to the driver threads
- benchmark: disable bucket histogram in stats by default
- default minimum capacity changed to 64MB

0.2.1
=====

#1: KeyBuffer class failed for serialized key size of 16 and multiples of 16

0.2
===

intial release on Maven Central


ohc-benchmark/src/main/sh/batch-bench.sh
#!/bin/bash
#
# This script has been used to execute a bunch of benchmarks on a single host.
# It varies some parameters to check cache performance using mutations of parameters.
#

#
# Formula to calculate size of a single hash entry:
# HEADER_SIZE + key_length + ROUND_UP_8(key_length) + value_length
#
#   HEADER_SIZE is a constant = 56
#   ROUND_UP_8 rounds up key_length to the next 8 byte boundary
#
# IMPORTANT: always add 8 in mind since the benchmark uses a 'long' in the key
#
# Example:
#  key_length = 5 (+8)
#  value_length = 10
#  that is: 56 + (8 + 5 + 3) + 10 = 82
#
# To calculate the number of entries for a capacity is thus just
#   CAPACITY / ENTRY_LEN
#
# This does not include the overhead required by memory management (jemalloc, OS native allocator).
#

GIG_8=8589934592
GIG_16=$((2*GIG_8))
GIG_64=$((8*GIG_8))
GIG_192=$((24*GIG_8))

SEGMENTS=64
SEGMENTS4=256

#
# entry sizes vs. capacity      key-len     value-size      entry-size      capacity    # of entries
#                               256         4096            4480            8G          ~2M
#                               256         16384           16704           8G          ~500k
#                               256         65536           65856           8G          ~130k
#                               256         131072          131392          8G          ~65k
#
#
PARAMS=""
#                   Duration    Capacity    Threads     Segments    Ratio   Key-Len     Value-Size  Read-Key-Dist   Write-Key-Dist
PARAMS="$PARAMS     300         $GIG_8      8           $SEGMENTS   .8      0           4096        2000000         2000000"
PARAMS="$PARAMS     300         $GIG_64     8           $SEGMENTS   .8      0           4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    8           $SEGMENTS   .8      0           4096        48000000        48000000"
PARAMS="$PARAMS     300         $GIG_8      30          $SEGMENTS   .8      0           4096        8000000         8000000"
PARAMS="$PARAMS     300         $GIG_64     30          $SEGMENTS   .8      0           4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    30          $SEGMENTS   .8      0           4096        48000000        48000000"
PARAMS="$PARAMS     300         $GIG_8      64          $SEGMENTS   .8      0           4096        8000000         8000000"
PARAMS="$PARAMS     300         $GIG_64     64          $SEGMENTS   .8      0           4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    64          $SEGMENTS   .8      0           4096        48000000        48000000"

PARAMS="$PARAMS     300         $GIG_8      8           $SEGMENTS   .8      256         4096        2000000         2000000"
PARAMS="$PARAMS     300         $GIG_64     8           $SEGMENTS   .8      256         4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    8           $SEGMENTS   .8      256         4096        48000000        48000000"
PARAMS="$PARAMS     300         $GIG_8      30          $SEGMENTS   .8      256         4096        8000000         8000000"
PARAMS="$PARAMS     300         $GIG_64     30          $SEGMENTS   .8      256         4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    30          $SEGMENTS   .8      256         4096        48000000        48000000"
PARAMS="$PARAMS     300         $GIG_8      64          $SEGMENTS   .8      256         4096        8000000         8000000"
PARAMS="$PARAMS     300         $GIG_64     64          $SEGMENTS   .8      256         4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    64          $SEGMENTS   .8      256         4096        48000000        48000000"

PARAMS="$PARAMS     300         $GIG_8      8           $SEGMENTS   .8      512         4096        2000000         2000000"
PARAMS="$PARAMS     300         $GIG_64     8           $SEGMENTS   .8      512         4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    8           $SEGMENTS   .8      512         4096        48000000        48000000"
PARAMS="$PARAMS     300         $GIG_8      30          $SEGMENTS   .8      512         4096        8000000         8000000"
PARAMS="$PARAMS     300         $GIG_64     30          $SEGMENTS   .8      512         4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    30          $SEGMENTS   .8      512         4096        48000000        48000000"
PARAMS="$PARAMS     300         $GIG_8      64          $SEGMENTS   .8      512         4096        8000000         8000000"
PARAMS="$PARAMS     300         $GIG_64     64          $SEGMENTS   .8      512         4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    64          $SEGMENTS   .8      512         4096        48000000        48000000"

PARAMS="$PARAMS     300         $GIG_8      8           $SEGMENTS   .8      0           16384       500000          500000"
PARAMS="$PARAMS     300         $GIG_64     8           $SEGMENTS   .8      0           16384       4000000         4000000"
PARAMS="$PARAMS     300         $GIG_192    8           $SEGMENTS   .8      0           16384       12000000        12000000"
PARAMS="$PARAMS     300         $GIG_8      30          $SEGMENTS   .8      0           16384       500000          500000"
PARAMS="$PARAMS     300         $GIG_64     30          $SEGMENTS   .8      0           16384       4000000         4000000"
PARAMS="$PARAMS     300         $GIG_192    30          $SEGMENTS   .8      0           16384       12000000        12000000"
PARAMS="$PARAMS     300         $GIG_8      64          $SEGMENTS   .8      0           16384       500000          500000"
PARAMS="$PARAMS     300         $GIG_64     64          $SEGMENTS   .8      0           16384       4000000         4000000"
PARAMS="$PARAMS     300         $GIG_192    64          $SEGMENTS   .8      0           16384       12000000        12000000"

PARAMS="$PARAMS     300         $GIG_8      8           $SEGMENTS   .8      0           65536       130000          130000"
PARAMS="$PARAMS     300         $GIG_64     8           $SEGMENTS   .8      0           65536       1000000         1000000"
PARAMS="$PARAMS     300         $GIG_192    8           $SEGMENTS   .8      0           65536       3000000         3000000"
PARAMS="$PARAMS     300         $GIG_8      30          $SEGMENTS   .8      0           65536       130000          130000"
PARAMS="$PARAMS     300         $GIG_64     30          $SEGMENTS   .8      0           65536       1000000         1000000"
PARAMS="$PARAMS     300         $GIG_192    30          $SEGMENTS   .8      0           65536       3000000         3000000"
PARAMS="$PARAMS     300         $GIG_8      64          $SEGMENTS   .8      0           65536       130000          130000"
PARAMS="$PARAMS     300         $GIG_64     64          $SEGMENTS   .8      0           65536       1000000         1000000"
PARAMS="$PARAMS     300         $GIG_192    64          $SEGMENTS   .8      0           65536       3000000         3000000"

#

PARAMS="$PARAMS     300         $GIG_8      8           $SEGMENTS   .5      0           4096        2000000         2000000"
PARAMS="$PARAMS     300         $GIG_64     8           $SEGMENTS   .5      0           4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    8           $SEGMENTS   .5      0           4096        48000000        48000000"
PARAMS="$PARAMS     300         $GIG_8      30          $SEGMENTS   .5      0           4096        8000000         8000000"
PARAMS="$PARAMS     300         $GIG_64     30          $SEGMENTS   .5      0           4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    30          $SEGMENTS   .5      0           4096        48000000        48000000"
PARAMS="$PARAMS     300         $GIG_8      64          $SEGMENTS   .5      0           4096        8000000         8000000"
PARAMS="$PARAMS     300         $GIG_64     64          $SEGMENTS   .5      0           4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    64          $SEGMENTS   .5      0           4096        48000000        48000000"

PARAMS="$PARAMS     300         $GIG_8      8           $SEGMENTS   .5      0           16384       500000          500000"
PARAMS="$PARAMS     300         $GIG_64     8           $SEGMENTS   .5      0           16384       4000000         4000000"
PARAMS="$PARAMS     300         $GIG_192    8           $SEGMENTS   .5      0           16384       12000000        12000000"
PARAMS="$PARAMS     300         $GIG_8      30          $SEGMENTS   .5      0           16384       500000          500000"
PARAMS="$PARAMS     300         $GIG_64     30          $SEGMENTS   .5      0           16384       4000000         4000000"
PARAMS="$PARAMS     300         $GIG_192    30          $SEGMENTS   .5      0           16384       12000000        12000000"
PARAMS="$PARAMS     300         $GIG_8      64          $SEGMENTS   .5      0           16384       500000          500000"
PARAMS="$PARAMS     300         $GIG_64     64          $SEGMENTS   .5      0           16384       4000000         4000000"
PARAMS="$PARAMS     300         $GIG_192    64          $SEGMENTS   .5      0           16384       12000000        12000000"

#

PARAMS="$PARAMS     300         $GIG_8      8           $SEGMENTS   .2      0           4096        2000000         2000000"
PARAMS="$PARAMS     300         $GIG_64     8           $SEGMENTS   .2      0           4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    8           $SEGMENTS   .2      0           4096        48000000        48000000"
PARAMS="$PARAMS     300         $GIG_8      30          $SEGMENTS   .2      0           4096        8000000         8000000"
PARAMS="$PARAMS     300         $GIG_64     30          $SEGMENTS   .2      0           4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    30          $SEGMENTS   .2      0           4096        48000000        48000000"
PARAMS="$PARAMS     300         $GIG_8      64          $SEGMENTS   .2      0           4096        8000000         8000000"
PARAMS="$PARAMS     300         $GIG_64     64          $SEGMENTS   .2      0           4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    64          $SEGMENTS   .2      0           4096        48000000        48000000"

PARAMS="$PARAMS     300         $GIG_8      8           $SEGMENTS   .2      0           16384       500000          500000"
PARAMS="$PARAMS     300         $GIG_64     8           $SEGMENTS   .2      0           16384       4000000         4000000"
PARAMS="$PARAMS     300         $GIG_192    8           $SEGMENTS   .2      0           16384       12000000        12000000"
PARAMS="$PARAMS     300         $GIG_8      30          $SEGMENTS   .2      0           16384       500000          500000"
PARAMS="$PARAMS     300         $GIG_64     30          $SEGMENTS   .2      0           16384       4000000         4000000"
PARAMS="$PARAMS     300         $GIG_192    30          $SEGMENTS   .2      0           16384       12000000        12000000"
PARAMS="$PARAMS     300         $GIG_8      64          $SEGMENTS   .2      0           16384       500000          500000"
PARAMS="$PARAMS     300         $GIG_64     64          $SEGMENTS   .2      0           16384       4000000         4000000"
PARAMS="$PARAMS     300         $GIG_192    64          $SEGMENTS   .2      0           16384       12000000        12000000"

#

PARAMS="$PARAMS     300         $GIG_8      8           $SEGMENTS4  .8      0           4096        2000000         2000000"
PARAMS="$PARAMS     300         $GIG_64     8           $SEGMENTS4  .8      0           4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    8           $SEGMENTS4  .8      0           4096        48000000        48000000"
PARAMS="$PARAMS     300         $GIG_8      30          $SEGMENTS4  .8      0           4096        8000000         8000000"
PARAMS="$PARAMS     300         $GIG_64     30          $SEGMENTS4  .8      0           4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    30          $SEGMENTS4  .8      0           4096        48000000        48000000"
PARAMS="$PARAMS     300         $GIG_8      64          $SEGMENTS4  .8      0           4096        8000000         8000000"
PARAMS="$PARAMS     300         $GIG_64     64          $SEGMENTS4  .8      0           4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    64          $SEGMENTS4  .8      0           4096        48000000        48000000"

PARAMS="$PARAMS     300         $GIG_8      8           $SEGMENTS4  .8      0           16384       500000          500000"
PARAMS="$PARAMS     300         $GIG_64     8           $SEGMENTS4  .8      0           16384       4000000         4000000"
PARAMS="$PARAMS     300         $GIG_192    8           $SEGMENTS4  .8      0           16384       12000000        12000000"
PARAMS="$PARAMS     300         $GIG_8      30          $SEGMENTS4  .8      0           16384       500000          500000"
PARAMS="$PARAMS     300         $GIG_64     30          $SEGMENTS4  .8      0           16384       4000000         4000000"
PARAMS="$PARAMS     300         $GIG_192    30          $SEGMENTS4  .8      0           16384       12000000        12000000"
PARAMS="$PARAMS     300         $GIG_8      64          $SEGMENTS4  .8      0           16384       500000          500000"
PARAMS="$PARAMS     300         $GIG_64     64          $SEGMENTS4  .8      0           16384       4000000         4000000"
PARAMS="$PARAMS     300         $GIG_192    64          $SEGMENTS4  .8      0           16384       12000000        12000000"

#

PARAMS="$PARAMS     300         $GIG_8      8           $SEGMENTS4  .2      0           4096        2000000         2000000"
PARAMS="$PARAMS     300         $GIG_64     8           $SEGMENTS4  .2      0           4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    8           $SEGMENTS4  .2      0           4096        48000000        48000000"
PARAMS="$PARAMS     300         $GIG_8      30          $SEGMENTS4  .2      0           4096        8000000         8000000"
PARAMS="$PARAMS     300         $GIG_64     30          $SEGMENTS4  .2      0           4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    30          $SEGMENTS4  .2      0           4096        48000000        48000000"
PARAMS="$PARAMS     300         $GIG_8      64          $SEGMENTS4  .2      0           4096        8000000         8000000"
PARAMS="$PARAMS     300         $GIG_64     64          $SEGMENTS4  .2      0           4096        16000000        16000000"
PARAMS="$PARAMS     300         $GIG_192    64          $SEGMENTS4  .2      0           4096        48000000        48000000"

PARAMS="$PARAMS     300         $GIG_8      8           $SEGMENTS4  .2      0           16384       500000          500000"
PARAMS="$PARAMS     300         $GIG_64     8           $SEGMENTS4  .2      0           16384       4000000         4000000"
PARAMS="$PARAMS     300         $GIG_192    8           $SEGMENTS4  .2      0           16384       12000000        12000000"
PARAMS="$PARAMS     300         $GIG_8      30          $SEGMENTS4  .2      0           16384       500000          500000"
PARAMS="$PARAMS     300         $GIG_64     30          $SEGMENTS4  .2      0           16384       4000000         4000000"
PARAMS="$PARAMS     300         $GIG_192    30          $SEGMENTS4  .2      0           16384       12000000        12000000"
PARAMS="$PARAMS     300         $GIG_8      64          $SEGMENTS4  .2      0           16384       500000          500000"
PARAMS="$PARAMS     300         $GIG_64     64          $SEGMENTS4  .2      0           16384       4000000         4000000"
PARAMS="$PARAMS     300         $GIG_192    64          $SEGMENTS4  .2      0           16384       12000000        12000000"

#

warm_up=15
cold_sleep=5
mode="exec"
j_exec=""
pre=""
while [ $# -ne 0 ] ; do
    case $1 in
        -warm)  warm_up=$2; shift;;
        -cold)  cold_sleep=$2; shift;;
        -count) mode="count";;
        -dry)   mode="dry";;
        -pre)   pre="$2 "; shift;;
        -X)     jvm_arg="$JVM_ARG $2"; shift;;
    esac
    shift
done

#

script_dir=`dirname $0`
base_dir="$script_dir/../../.."
version=`grep '<version>' -m 1 pom.xml | sed 's/.*<version>\(.*\)<\/version>.*/\1/'`
jar=$base_dir/target/ohc-benchmark-$version.jar
if [ ! -f $jar ] ; then
    echo "Executable jar file $jar does not exist" > /dev/stderr
    exit 1
fi
java -jar $jar -h > /dev/null
if [ ! $? ] ; then
    echo "Cannot execute ohc-benchmark jar file $jar" > /dev/stderr
    exit 1
fi

time_cmd=
if [ `which gtime 2> /dev/null` ] ; then
    time_cmd=gtime
else
    time_cmd=time
fi

t_exec="$time_cmd -f %e;%S;%U;%P;%M;%K;\"%C\" "
j_exec="$j_exec java $jvm_arg -jar $jar "

capacity=
threads=
segments=
ratio=
key_len=
value_size=
read_key_dist=
write_key_dist=
p_num=1
count=0
seconds=0

log_dir=./batch-bench/`date "+%Y-%m-%d-%H-%M-%S"`
echo "Log directory for this batch is $log_dir"
mkdir -p $log_dir
if [ ! $? ] ; then
    echo "Cannot create $log_dir" > /dev/stderr
    exit 1
fi

exec >  >(tee -a $log_dir/exec.log)
exec 2> >(tee -a $log_dir/exec.log >&2)

for p in $PARAMS ; do
    case $p_num in
        1)  duration=$p;        p_num=$((p_num+1));;
        2)  capacity=$p;        p_num=$((p_num+1));;
        3)  threads=$p;         p_num=$((p_num+1));;
        4)  segments=$p;        p_num=$((p_num+1));;
        5)  ratio=$p;           p_num=$((p_num+1));;
        6)  key_len=$p;         p_num=$((p_num+1));;
        7)  value_size=$p;      p_num=$((p_num+1));;
        8)  read_key_dist=$p;   p_num=$((p_num+1));;
        9)  write_key_dist=$p
            p_num=1
            count=$((count+1))
            seconds=$((seconds+duration+warm_up+cold_sleep))
            name=`printf "bench-%03d" $count`"-c$capacity-kl$key_len-r$ratio-sc$segments-t$threads-rkd$read_key_dist-wkd$write_key_dist-vs$value_size"

            export $pre
            j_run="$t_exec -o $log_dir/$name-time.csv"
            j_run="$j_run $j_exec -cap $capacity -d $duration -kl $key_len -r $ratio -sc $segments -t $threads -wu \"$warm_up,$cold_sleep\""
            j_run="$j_run -rkd gaussian(1..$read_key_dist,2)"
            j_run="$j_run -wkd gaussian(1..$write_key_dist,2)"
            j_run="$j_run -vs gaussian(1024..$value_size,2)"
            j_run="$j_run -csv $log_dir/$name.csv"

            case $mode in
                "count")    ;;
                "dry")      echo "   would execute:   $j_run"
                            ;;
                "exec")
                            $j_run > $log_dir/$name.log
                            ;;
            esac

            ;;
    esac
done

echo "$count executions , runtime of $seconds seconds ($((seconds/60)) minutes)"
